### PR TITLE
Scale cover image and add a background where necessary to better fit the screen

### PR DESCRIPTION
**Pull request by zwim**
This should fix https://github.com/koreader/koreader/issues/6876

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/6884)

### CODE REVIEW

**Comment by poire-z** [comment]
Given that these options for the regular screensaver are saved in G_reader_settings, and so easily accessible from your plugin, you could re-use their values:
https://github.com/koreader/koreader/blob/6cc0ab8dc9da1549f6a6d1abdb8fe190b302faba/frontend/ui/screensaver.lua#L168-L178
especially the white background (so people with a white plastic device can get white stripes).

**Comment by zwim** [comment]
Yes, we could add a menu-entry "Scale book cover" with the entries:
[ ] Black background
[ ] White background
[ ] No background

**Comment by poire-z** [comment]
Why not if you're up for that (my suggestion was just to avoid that menu work :)

File: plugins/coverimage.koplugin/main.lua [hunk 1]
```diff
@@ -67,9 +71,40 @@ end
...
+            end
+
+            local scaled_w, scaled_h = i_w * scale_factor, i_h * scale_factor
+            local cover_image = RenderImage:scaleBlitBuffer(cover_image, scaled_w, scaled_h)
+
+            -- new buffer with screen dimensions,
+            local image = Blitbuffer.new( s_w, s_h, cover_image:getType() )
+            logger.err("xxxxxxxxxxxxxxxxxxxxxxxx " .. self.cover_image_background)
+            if self.cover_image_background == "black" then
+                image:fill(Blitbuffer.COLOR_BLACK)
```
Comment (NiLuJe) [inline plugins/coverimage.koplugin/main.lua:92]: BBs are zero-initialized, so the explicit zero-fill is technically redundant.

(I'd still keep it just to avoid aking the code more puzzling in that case, this is just me throwing in a "fun fact" here ^^).

File: plugins/coverimage.koplugin/main.lua [hunk 2]
```diff
@@ -67,9 +71,39 @@ end
...
+            local i_w, i_h = cover_image:getWidth(), cover_image:getHeight()
+            local scale_factor = math.min(s_w / i_w, s_h / i_h)
+
+            if self.cover_image_background == "none" or scale_factor == 1 then
+                cover_image:writePNG(self.cover_image_path, false)
+                return
+            end
+
+            local scaled_w, scaled_h = i_w * scale_factor, i_h * scale_factor
+            local cover_image = RenderImage:scaleBlitBuffer(cover_image, scaled_w, scaled_h)
```
Comment (Galunid) [inline plugins/coverimage.koplugin/main.lua:86]: Do those work properly? In https://github.com/koreader/koreader/pull/6709 using "float" in blitbuffer caused image to be "projected".

**Review by Galunid — COMMENTED** [review]
Other than those two comments it looks good to me.

File: plugins/coverimage.koplugin/main.lua [hunk 3]
```diff
@@ -36,10 +38,12 @@ local CoverImage = WidgetContainer:new{
 
 function CoverImage:init()
     self.cover_image_path = G_reader_settings:readSetting("cover_image_path") or "cover.png"
+    self.cover_image_background = G_reader_settings:readSetting("cover_image_background") or "none"
```
Comment (Galunid) [inline plugins/coverimage.koplugin/main.lua:41]: It should probably be enabled by default. This shouldn't take too much time and it would ensure it works out of the box on all devices.

File: plugins/coverimage.koplugin/main.lua [hunk 4]
```diff
@@ -67,9 +71,39 @@ end
...
+                return
+            end
+
+            local scaled_w, scaled_h = i_w * scale_factor, i_h * scale_factor
+            local cover_image = RenderImage:scaleBlitBuffer(cover_image, scaled_w, scaled_h)
+
+            -- new buffer with screen dimensions,
+            local image = Blitbuffer.new( s_w, s_h, cover_image:getType() )
+            if self.cover_image_background == "black" then
+--                image:fill(Blitbuffer.COLOR_BLACK) -- not necessary, as BB are zero-initialized
```
Comment (poire-z) [inline plugins/coverimage.koplugin/main.lua:91]: Please keep the `--` indented at the code level.
Also, no CI check seems to have run, but it will probably complain about the empty `then elseif`. When that's needed, one have to make it non empty with:
```lua
if something then
    do end -- luacheck: ignore 541
elseif
    other stuff
end
```
But I would probably leave the image.fill even if it's redundant. Or just chec for white and gray, and mention as a comment that black doesn't need it.

File: plugins/coverimage.koplugin/main.lua (hunk 2 above)
Comment (zwim) [reply plugins/coverimage.koplugin/main.lua:86]: For me it seems, that the images are clear, but a `math.floor` can not hurt.

File: plugins/coverimage.koplugin/main.lua (hunk 3 above)
Comment (zwim) [reply plugins/coverimage.koplugin/main.lua:41]: OK, set it to black as this is the fastest method and it fits best to my Tolino

File: plugins/coverimage.koplugin/main.lua (hunk 4 above)
Comment (zwim) [reply plugins/coverimage.koplugin/main.lua:91]: Thanks for the hint. I will mention the behavior  on `Blitbuffer.new`.

**Review by Galunid — APPROVED** [review]

**Comment by zwim** [comment]
Wouldn't it be better to save the background color globally and not for each doc?
The background color was introduced for people with white plastic cover, this does not depend on the book, but on the device.

**Comment by poire-z** [comment]
> Wouldn't it be better to save the background color globally and not for each doc?

Indeed.
But you read it from G_reader_settings on line 41.
And you save it in the doc_settings in the menu, which looks like an error :)

**Comment by zwim** [comment]
> But you read it from G_reader_settings on line 41.
> And you save it in the doc_settings in the menu, which looks like an error :)

Yes, I have mixed them up. But now it should be corrected, only G_reader_settings is used now.

File: plugins/coverimage.koplugin/main.lua [hunk 5]
```diff
@@ -193,6 +227,70 @@ function CoverImage:addToMainMenu(menu_items)
...
+                        text = _("White background"),
+                        checked_func = function()
+                            return self.cover_image_background == "white"
+                        end,
+                        callback = function()
+                            if self.cover_image_background == "white" then
+                                self.cover_image_background = "none"
+                            else
+                                self.cover_image_background = "white"
+                            end
```
Comment (Galunid) [inline plugins/coverimage.koplugin/main.lua:268]: It could be rewritten as something like
```lua
self.cover_image_background = self.cover_image_background == "white" and "none" or "white"
```

File: plugins/coverimage.koplugin/main.lua [hunk 6]
```diff
@@ -193,6 +227,70 @@ function CoverImage:addToMainMenu(menu_items)
...
+                        text = _("Gray background"),
+                        checked_func = function()
+                            return self.cover_image_background == "gray"
+                        end,
+                        callback = function()
+                            if self.cover_image_background == "gray" then
+                                self.cover_image_background = "none"
+                            else
+                                self.cover_image_background = "gray"
+                            end
```
Comment (Galunid) [inline plugins/coverimage.koplugin/main.lua:285]: Same as above

File: plugins/coverimage.koplugin/main.lua [hunk 7]
```diff
@@ -193,6 +227,70 @@ function CoverImage:addToMainMenu(menu_items)
                     end
                 end,
             },
+            -- menu entry: scale book cover
+            {
+                text = _("Scale book cover"),
+                help_text = _("If disabled, the cover will have to be scaled by the FW.\nDisabling is the fastest method, but doesn't work on all devices."),
+                checked_func = function()
+                    return self.ui and self.ui.doc_settings and self.cover_image_background ~= "none"
```
Comment (Galunid) [inline plugins/coverimage.koplugin/main.lua:235]: No need to check for `self.ui` and `self.ui.doc_settings` here now :)

Comment (zwim) [reply plugins/coverimage.koplugin/main.lua]: Right!

**Comment by Galunid** [comment]
Off topic, why isn't there CI on this PR?

**Comment by Frenzie** [comment]
I think CircleCI quit running for people who aren't on it or something. It's quite annoying. >_<

**Comment by sndwichm4n** [comment]
> CircleCI

What is this?

**Comment by Frenzie** [comment]
See https://app.circleci.com/pipelines/github/koreader/koreader/1429/workflows/6e51085b-43cd-4958-ac16-f765ed357056/jobs/5738 for a sample run. It runs a variety of automated tests we set up to ensure certain things don't break and/or meet a certain level of quality.

**Comment by sndwichm4n** [comment]
Very interesting, thanks.
> Good Job
> 498 tests are passing. 
:D

**Comment by poire-z** [comment]
> I think CircleCI quit running for people who aren't on it or something. It's quite annoying

Indeed :( We (members, any logged in user?) can still re-run "unauthorized" builds - which I just did for our opened PRs that were missing it.

File: plugins/coverimage.koplugin/main.lua [hunk 8]
```diff
@@ -193,6 +228,58 @@ function CoverImage:addToMainMenu(menu_items)
                     end
                 end,
             },
+            -- menu entry: scale book cover
+            {
+                text = _("Scale book cover"),
+                help_text = _("If disabled, the cover will have to be scaled by the FW.\nDisabling is the fastest method, but doesn't work on all devices."),
```
Comment (poire-z) [inline plugins/coverimage.koplugin/main.lua:234]: Haven't seen how this is presented, but it seems a bit strange:
`text = _("Scale book cover"),` hints it might be a toggle, or some scaling options
`sub_item_table ` presents "black|white|grey background" options, which seem totally unrelated to scaling :) 
`help_text` is then not helpfull, mentionning "disabling" which one can't find anywhere :/

Comment (zwim) [reply plugins/coverimage.koplugin/main.lua:234]: Should I rename `Scale book cover` to `Scale book cover and add background` and
the help text to:   _("If a background color is selected, the image will be scaled with its original aspect ratio and a solid backround to fit the device screen.\nIf no background color is selected, the cover is saved in the original aspect ratio and it has to be scaled by the FW.\nThe latter is the fastest method, but doesn't work on all devices.")?

Comment (poire-z) [reply plugins/coverimage.koplugin/main.lua:234]: We use the word "Stretch" in the regular screensaver - it sounds clearer than "Scale" in expressing that the image will be distorted :) 
May be you should have 4 items instead of 3, mutually exclusive: `[ ] Strech to fit screen`, with always one checked.
Your main menu title could then be `Stretching and background`.

**Comment by zwim** [comment]
@pazos: I have change the sub menu the text. 
When pushing the commit 43651c0 I got the message that I should do a `git pull` first, and now I have a strange commit 29dcc6b in here. If this does any harm, how can I remove it?

File: plugins/coverimage.koplugin/main.lua [hunk 9]
```diff
@@ -270,7 +267,20 @@ function CoverImage:addToMainMenu(menu_items)
...
+                            G_reader_settings:saveSetting("cover_image_background", self.cover_image_background)
+                            if self.enabled then
+                                self:createCoverImage(self.ui.doc_settings)
+                            end
+                        end
+                    },
+                    {
+                        text = _("Stretch to fit screen"),
+                        checked_func = function()
+                            return self.cover_image_background == "none"
```
Comment (poire-z) [inline plugins/coverimage.koplugin/main.lua:280]: Is their a 5th alternative behaviour when `self.cover_image_background` is `nil` ?
If yes, it would be nice it appears as the first option "Save cover as-is".
If no, one of these 4 should handle `self.cover_image_background == nil` in the checked_func.

**Comment by poire-z** [comment]
> now I have a strange commit 29dcc6b in here. If this does any harm, how can I remove it?

I'd say: don't worry about it :) as long as it looks good in the global `Files changed` tab.
(We'll squash all your commits when merging, so we won't see that bogus commit.)

File: plugins/coverimage.koplugin/main.lua (hunk 9 above)
Comment (zwim) [reply plugins/coverimage.koplugin/main.lua:280]: We should not have a 5th alternative, because actually we do not stretch the image to fullscreen, we save it as is. The firmware of the device can take this and scale, stretch or do whatever the device does. 
So actually the 4th menu entry should be named better, something the way: "Save image as is and let the firmware deal with stretching" (in better word of course).

`self.cover_image_background` cannot be `nil` because we initialize it in line 43 with "black" if it is nil.

Comment (zwim) [reply plugins/coverimage.koplugin/main.lua:280]: Maybe we could name it "Save the original image" and make a help_text with the info "The firmware can stretch the image to fit the screen"

Comment (poire-z) [reply plugins/coverimage.koplugin/main.lua:280]:
```
white
gray
black
stretch
as-is
```
That makes 5. Which one would you remove?

(Or do you mean you don't really need stretch ? And what I thought was better named stetch was actually just doing scaling to screen size, preservering aspect ratio? So, as-is just scaling to screen size, keep aspect ratio - and we can actually keep it doing that while still naming it as-is, for simplicity).

Let's keep using "scale" when the aspect ratio is preserved (and "stretch" only when it is not).

Comment (zwim) [reply plugins/coverimage.koplugin/main.lua:280]: We have:
black -> scale and add black background
white -> scale and add white background
gray -> scale and add gray background
as-is  -> just save 

Until now, the stretch option is not implemented in the plugin, because it is not needed. The firmware takes an image and stretches it to fully fit the screen (which does not work very good on Poke2, but works on Tolino).

Of course, we can add a fifth option `stretch`, if you think that's necessary.

I am not really satisfied with the naming of the menu options, wouldn't it  be better to name the thing: `Size and background` and the submenus like
```
[ ] Original size
[ ] Scale, black background
[ ] Scale, white background
[ ] Scale, gray background
```

Comment (poire-z) [reply plugins/coverimage.koplugin/main.lua:280]: OK, I'm fine with that proposal (no need for stretch if it's the default behaviour on some FW, and otherwise the least interesting option as the others are better).

**Comment by poire-z** [comment]
Ready to merge ?
@Frenzie: fine with the wording ?

**Comment by zwim** [comment]
@poire-z: From my side it is ready.
Let's see, what magic @Frenzie does with the wording ;-)

File: plugins/coverimage.koplugin/main.lua (hunk 9 above)
Comment (Frenzie) [reply plugins/coverimage.koplugin/main.lua:280]: That looks fine to me. :+1:

**Review by Frenzie — APPROVED** [review]
